core/shared/src/main/scala-dotty/zio/VersionSpecific.scala

```diff
@@ -1,15 +1,29 @@
 package zio
 
+import scala.reflect.ClassTag
+
 private[zio] trait VersionSpecific {
 
-  type Tagged[A] = scala.reflect.ClassTag[A]
+  // Workaround for disabled summon for ClassTag[Nothing]
+  // due to https://github.com/lampepfl/dotty/issues/1730
+  final class ClassTagBox[A](val cls: ClassTag[A])
+
+  object ClassTagBox extends LowPrio {
+    implicit def fromClassTag[T](implicit tag: ClassTag[T]): ClassTagBox[T] = new ClassTagBox(tag)
+  }
+
+  abstract class LowPrio {
+    implicit val classTagNothing: ClassTagBox[Nothing] = new ClassTagBox(ClassTag(classOf[Nothing]))
```

Comment:
Maybe next we'll disallow `classOf[Nothing]` to prevent this kind of misuse.

Reply:
To be 100% clear, you should not use `classOf[Nothing]`:
- It's not going to work properly (`taggedIsSubtype(taggedTagType(classTagNothing), anyOtherTag)` will always return false, when it should always return true).
- It leads to the same kind of soundness hole we had previously, so we'll have to disallow it and your code will stop compiling.

Reply:
@smarter This wouldn't make much sense, in Scala 2 `classOf` cannot fail, even for class-less types like abstract type members. There is _always_ an available erasure for any concrete type.

Reply:
I'm not sure why you insist this is misuse. `Nothing` is just a special case we're not particuralily interested in, it just has to be supported becouse it is used all over ZIO as a marker that an effect can't fail. I'll try to give it some more thought tomorrow, or maybe you could explain in more details why this is wrong?

Reply:
> This wouldn't make much sense, in Scala 2 classOf cannot fail, even for class-less types like abstract type members. 

That doesn't seem to be true, I tried 2.13.1:
```scala
scala> def foo[T] = classOf[T]
                            ^
       error: class type required but T found

scala> class Foo { type A; classOf[A] }
                                   ^
       error: class type required but Foo.this.A found
```


> I'll try to give it some more thought tomorrow, or maybe you could explain in more details why this is wrong

`classOf[Nothing]` returns `scala.runtime.Nothing$` a dummy class that is never instantiated and is not a subclass or superclass of anything else. You told me you needed this ClassTag for Method, I looked at Method and saw:
```scala
      taggedIsSubtype(self.inputTag, that.inputTag) &&
      taggedIsSubtype(self.errorTag, that.errorTag) &&
      taggedIsSubtype(self.outputTag, that.outputTag)
```
and the definition of `taggedIsSubtype` is:
```scala
  private[zio] def taggedIsSubtype[A, B](left: TagType, right: TagType): Boolean =
    right.isAssignableFrom(left)
```
**This will never return true when left is classOf[Nothing], even though Nothing is a subtype of every other type** so this is wrong (it is wrong in many other ways too, there are many situations where subtyping and subclassing are different).

Reply:
@smarter 
> That doesn't seem to be true, I tried 2.13.1:

It succeeds for all stable paths. (except type projections)

> classOf[Nothing] returns scala.runtime.Nothing$ a dummy class that is never instantiated

It's used in Scala 2 in erasure of methods returning Nothing. Also `.asInstanceOf[Nothing]` will try to cast to that class - it's not total compiler fiction, Nothing does refer to a material class currently.

```
def erasure: Nothing = ???
public scala.runtime.Nothing$ erasure()
```

Whenever working with java reflection it still makes sense to e.g. check method return type against `classOf[Nothing]`, as long as the latter exists in bytecode, so I disagree that it would make any sense to remove it.
For `ClassTag`, `Array` can use an opaque type over ClassTag that disallows bottoms without changing ClassTag or breaking binary compatibility, I'll leave it for you to judge whether that's worth the indirection, but I do think that a soundness problem with `Array` should lead to changes in `Array`, not to changes in `ClassTag`, a thing that's only tangentially related to `Array` and makes sense for as long as `scala.runtime.Nothing$` actually exists in bytecode.

> This will never return true when left is classOf[Nothing], even though Nothing is a subtype of every other type so this is wrong (it is wrong in many other ways too, there are many situations where subtyping and subclassing are different).

This is true, but there are also other cases where `isAssignableFrom` will fail when in Scala it would succeed either at runtime or at compile-time (`int`vs.`Int`vs.`Integer`). (@ioleo Clearly you'll need to add special cases in `taggedIsSubtype` for this to work, yep)

Reply:
> It succeeds for all stable paths. (except type projections)

I don't understand what you mean by that.

> Whenever working with java reflection it still makes sense to e.g. check method return type against classOf[Nothing], as long as the latter exists in bytecode, so I disagree that it would make any sense to remove it.

In general, you can't rely on `classOf` to give you a type that matches exactly the result type of a method, for example a method that returns `(Int, Int)` in source code will return `scala.Tuple2$mcII$sp` in bytecode, but `classOf[(Int, Int)` returns `scala.Tuple2`.

> I do think that a soundness problem with Array should lead to changes in Array, not to changes in ClassTag, a thing that's only tangentially related to Array

At this point, I think that making array is the only justified usage of ClassTag, it's possible to use it to avoid unchecked warning, but this is actually unsound and will have to be removed: https://github.com/lampepfl/dotty/issues/7554. Otherwise it seems to be mostly used incorrectly (like in this PR, where it is mistaken as a way to do subtype checks at runtime).

> This is true, but there are also other cases where isAssignableFrom will fail when in Scala it would succeed either at runtime or at compile-time (intvs.Intvs.Integer

Exactly, you can't do subtype checks at runtime using ClassTag at all, so it doesn't matter if ClassTag[Nothing] exists or not, this is fundamentally the wrong approach.

> Clearly you'll need to add special cases in taggedIsSubtype for this to work

No amount of special cases that you add will ever make this correct in general.

Reply:
@smarter 
> I don't understand what you mean by that.

I was wrong, `classOf` only works for class types and intersections, 

> Exactly, you can't do subtype checks at runtime using ClassTag at all, so it doesn't matter if ClassTag[Nothing] exists or not, this is fundamentally the wrong approach.

> At this point, I think that making array is the only justified usage of ClassTag, it's possible to use it to avoid unchecked warning

Fine, but the ecosystem uses ClassTag for a lot besides that, e.g. printing the class name is `pureconfig` - `ClassTag[Nothing]` is completely correct for printing.

> No amount of special cases that you add will ever make this correct in general.

It can make it correct for a narrow amount of well known cases - such as unparameterized concrete class types, which is all that's necesasry until [`izumi-reflect`](https://github.com/zio/izumi-reflect) for dotty is ready and will be able handle a larger amount of known cases.

Reply:
> Fine, but the ecosystem uses ClassTag for a lot besides that

I think the ecosystem will have to switch to other typeclasses depending on their usecase. Either making their own (like Typeable in shapeless: https://github.com/milessabin/shapeless/blob/shapeless-3/core/src/main/scala/shapeless/typeable.scala, or reusing the stuff in dotty like [scala.quoted.Type](http://dotty.epfl.ch/docs/reference/metaprogramming/macros.html) (currently only usable from inside an inline def, but maybe that limitation can be lifted for globally accessible types).